dhcp-relay/dhcp_kern_xdp.c

```diff
@@ -49,77 +56,162 @@ struct {
 	__uint(max_entries, 16384);
 } client_vlans SEC(".maps");
 
+//static int memcpy_var(void *to, void *from, __u8 len) {
+//	__u8 *t8 = to, *f8 = from;
+//	int i;
+//
+//	if (len > RAI_OPTION_LEN) {
+//		return -1;
+//	}
+//
+//	for (i = 0; i < len; i++) {
+//
+//		if (i > RAI_OPTION_LEN) {
+//			return -1;
+//		}
+//
+//		*t8++ = *f8++;
+//
+//	}
+//
+//
+//	if (i == RAI_OPTION_LEN)
+//		return -1;
+//
+//	return 0;
+//
+//}
+
+int u16_to_ascii(struct sub_option *opt, __u8 offset, __u16 num) {
+
+	if (opt == NULL)
+		return -1;
+
+	if (offset > RAI_OPTION_LEN)
+		return -1;
+
+	if (offset < U16_ASCII_LEN)
+		return -1;
+
+	__u8 i;
+	//#pragma unroll U16_ASCII_LEN
+	for (i = offset - 1; i > 0; i--) {
+
+		if (i == 0)
+			break;
+
+		opt->val[i] = (__u8) (num % 10) + '0';
+		num /= 10;
+		if (num == 0)
+			break;
+
+	}
+
+	if (i > 0) {
+		opt->val[--i] = '.';
+	}
+
+	if (i > RAI_OPTION_LEN)
+		return -1;
+
+	return i;
+
+}
+
+int str_len(struct dev_name *dev) {
+
+	if (dev == NULL)
+		return -1;
+
+	__u8 i = 0;
+	for (i = 0; i < RAI_OPTION_LEN; i++)
+		if (dev->name[i] == 0)
+			break;
+
+	return i;
+
+}
+
+int copy_dev_name(struct sub_option *dest, __u8 offset, struct dev_name *dev) {
+
+	if (dest == NULL)
+		return -1;
+
+	if (dev == NULL)
+		return -1;
+
+	__u8 i;
+
+	/* Copy device name and left-align VLAN part*/
+#pragma unroll RAI_OPTION_LEN
+	for (i = 0; i < RAI_OPTION_LEN; i++) {
+
+		if (i < dev->len) {
+			/* Copy device name */
+			dest->val[i] = dev->name[i];
+		} else if (offset < RAI_OPTION_LEN) {
+			/* Move VLAN part (all bytes from offset and up) */
+			dest->val[i] = dest->val[offset];
+			dest->val[offset++] = 0;
+		}
```

Comment:
This weird double-construction supposes a particular construction of the buffer being written to, which is hard to understand and makes the code less readable (and less reusable as well). Instead, how about turning this function into one that actually does what the name applies, like:

```c
static int copy_dev_name(struct sub_option *dest, __u8 wr_offset,
			 struct dev_name *dev, __u8 rd_offset) {
	int i;

	if (dest == NULL || dev == NULL || wr_offset > sizeof(dest->val) -1)
		return -1;

	for (i = 0; i < sizeof(dest->val) - wr_offset; i++) {
		if (i + rd_offset > sizeof(dev->name) ||
		    !dev->name[i + rd_offset])
			break;

		dest->val[i + wr_offset] = dev->name[i + rd_offset];
	}

 	return i;
}
```

and calling it like:

```c
	len = copy_dev_name(&option.circuit_id, 0, dev, 0);
	if (len < 0)
		return len;

	len = copy_dev_name(&option.circuit_id, len, &buf, vlan_offset);
	if (len < 0)
		return len;
```

where `buf` is a separate `struct dev_name` allocated on the stack, into which the VLAN tags are stringified.